apps/server/src/utils/upload.rs

```diff
@@ -1,59 +1,169 @@
+/// the provided `is_valid_content_type` function.
+///
+/// Optionally, an `expected_type_name` can be provided so that the error message can specify what
+/// type was expected.
+async fn validate_and_load_upload(
+	upload: &mut Multipart,
+	max_size: Option<usize>,
+	is_valid_content_type: impl Fn(&ContentType) -> bool,
+	expected_type_name: Option<&str>,
 ) -> APIResult<(ContentType, Vec<u8>)> {
 	let field = upload.next_field().await?.ok_or_else(|| {
 		APIError::BadRequest(String::from("No file provided in multipart"))
 	})?;
-
 	let raw_content_type = field.content_type().map(ContentType::from);
 
-	let bytes = field.bytes().await?;
-	let file_size = bytes.len();
+	// Load bytes of uploaded file
+	let (bytes, file_size) = load_field_up_to_size(field, max_size).await?;
 
-	if bytes.is_empty() || bytes.len() < 5 {
-		return Err(APIError::BadRequest("Uploaded file is empty".to_string()));
-	}
-
-	let mut magic_bytes = vec![0; 5];
-	magic_bytes.copy_from_slice(&bytes[..5]);
-
-	let inferred_content_type = ContentType::from_bytes(&magic_bytes);
+	// Use first 5 bytes to infer content type.
+	// See: https://en.wikipedia.org/wiki/List_of_file_signatures
+	let magic_bytes = &bytes[..5];
+	let inferred_content_type = ContentType::from_bytes(magic_bytes);
 
 	let content_type = match (raw_content_type, inferred_content_type) {
-		(Some(provided), inferred) if !provided.is_image() && !inferred.is_image() => {
-			Err(APIError::BadRequest(
-				"Uploaded file is not an image".to_string(),
-			))
+		(Some(provided), inferred)
+			if !is_valid_content_type(&provided) && !is_valid_content_type(&inferred) =>
+		{
+			Err(validation_err(expected_type_name))
 		},
 		(Some(provided), inferred) => {
 			if provided != inferred {
 				tracing::warn!(?provided, ?inferred, "Content type mismatch");
 			}
-			let content_type = (inferred.is_image())
-				.then_some(inferred)
-				.or_else(|| provided.is_image().then_some(provided))
-				.ok_or_else(|| {
-					APIError::BadRequest("Uploaded file is not an image".to_string())
-				})?;
+			let content_type = if is_valid_content_type(&inferred) {
+				inferred
+			} else if is_valid_content_type(&provided) {
+				provided
+			} else {
+				return Err(validation_err(expected_type_name));
+			};
 			Ok(content_type)
 		},
-		(None, inferred) => (inferred.is_image()).then_some(inferred).ok_or_else(|| {
-			APIError::BadRequest("Uploaded file is not an image".to_string())
-		}),
+		(None, inferred) => {
+			if is_valid_content_type(&inferred) {
+				Ok(inferred)
+			} else {
+				Err(validation_err(expected_type_name))
+			}
+		},
 	}?;
 
-	tracing::trace!(?content_type, file_size, "Validated image upload");
-
+	tracing::trace!(?content_type, file_size, "Validated upload");
 	Ok((content_type, bytes.to_vec()))
 }
 
+/// Load up to `max_size` bytes of a field (erroring if `max_size` is exceeded).
+/// Returns the loaded bytes as [Vec<u8>] and the total bytes as [usize].
+async fn load_field_up_to_size(
```

Comment:
I'll have to double check the docs tomorrow, but I believe the `DefaultBodyLimit` middleware will enforce the limit for us. I'm not opposed to having extra safeguards to enforce it, but wanted to at least note it

Reply:
My thinking here was:

1) It seems better for the max_size to be enforced when this function is called, in case it is ever reused for something else in the future. I don't think the overhead is too notable, and theoretically many of these lines shouldn't run at all since the `DefaultBodyLimit` should immediately catch the request and return an error without running the body.
2) Looking at this documentation for stream size info, it looks like it's possible to send requests that are structured to deliver a large payload without signaling the size accurately in the headers. I don't know this for sure, but it might also prevent `DefaultBodyLimit` from catching the request. In this case, these lines should prevent users from downloading a massive image despite their set limits.

Reply:
Yeah like I said I'm not opposed to the extra safety, I agree that they shouldn't even hit for files which are too large anyways if the middleware does its job

> it looks like it's possible to send requests that are structured to deliver a large payload without signaling the size accurately in the headers

👀 I'd be curious to see if we can breal/trick the middleware at some point